dataplane/base.h

```diff
@@ -22,13 +22,33 @@ class PortMapper
 	uint16_t ports_count_ = 0;
 	tPortId dpdk_ports_[std::numeric_limits<tPortId>::max() + 1]; // logical to dpdk
 	tPortId logical_ports_[std::numeric_limits<tPortId>::max() + 1]; // dpdk to logical
```

Comment:
This class maps ports, but instead of using `std::unordered_map` we define two arrays with maximum size, and using `INVALID_PORT_ID` to mark absence of mapped key... Thus making this class 256 kilobytes in size? I mean, this _can_ be faster **if** we will have room for extra 256 kb in cache, which I doubt.

`std::unordered_map` is not a particularly good example, ideally we need some flat map, like `boost::flat_map` or any other open-source map implementation optimized for ints, like this, for example: https://github.com/Tessil/robin-map (MIT license)

But anyway, even an `std::unordered_map` is probably faster and certainly more convenient and readable than the current implementation

Reply:
The unordered map from stl has the downside of being very cache-unfriendly due to the guarantee that you can obtain a pointer to an object in the map, and this pointer remains valid even after rehashing or insertion/deletion of other elements. This basically necessitates a node-based implementation with things like linked lists.

However, I can see how maps can be faster if the number of keys is significantly smaller than 65535 and if we'll use different hash map implementation which don't have this requirement. Because since the memory footprint might be much smaller than with two 64KB arrays, it could fit within an L1 cache, for example.

But of course to make a definitive statement we would need to measure the performance in our specific situation